### ISSUE TITLE
Sanity check - overfit on single video

### ISSUE THREAD

**Issue by bredban323** [Rudrabha/Wav2Lip#336]
I was wondering if the authors (or anyone else for that matter) were successful training W2L from scratch on a single video? 

Theoretically, it should be possible to overfit the network on a single video unless the capacity of the network is insufficient. Essentially, I wanted to run a sanity check by training the generator and visual quality discriminator with a short 10 - 60 second video and its corresponding audio then check if the network is able to reproduce the original video on inference. So far, this endeavour has been unsuccessful (see additional information below).

Using `preprocess.py `for preprocessing, `hq_wav2lip_train.py `for training and `infernence.py `for inference, I find that the final output video does not move the lips corresponding to the audio. Furthermore, there are many other artifacts in the lip shape produced. 

Additional information:
- Video was of Charlie Chaplin dictator speech with original English audio (similar to W2L demo video)
- Trained for 2000 epochs
- Similar to what others have reported, after a few hundred epochs, discriminator real loss drops to 0 and fake loss goes to 100
- I made no modifications to the original code. All scripts and hparams are run as is.
- Sync loss jumps to 50-100 at times but generally stays around 0.69-1.x
- I tested the pretrained lip sync expert and the synchronization it gives on videos is random. I use as much of the original `color_syncnet_train.py `as possible and make no changes to dataloader. Furthermore, my example script for testing the lip sync expert is shown below. It is unable to adequately predict whether videos are in-sync or out-of-sync. 
- 

```
import argparse
from os.path import dirname, join, basename, isfile
import os, random, cv2, argparse

import numpy as np
from glob import glob

from hparams import hparams, get_image_list
from models import SyncNet_color as SyncNet
import audio

import torch
from torch import nn
from torch.utils import data as data_utils

parser = argparse.ArgumentParser(description='Code to train the Wav2Lip model WITH the visual quality discriminator')

parser.add_argument("--data_root", help="Root folder of the preprocessed dataset", required=True)
parser.add_argument('--syncnet_checkpoint_path', help='Load the pre-trained Expert discriminator', required=True,
                    type=str)

args = parser.parse_args()

use_cuda = torch.cuda.is_available()
print('use_cuda: {}'.format(use_cuda))
device = torch.device("cuda" if use_cuda else "cpu")
syncnet = SyncNet().to(device)

for p in syncnet.parameters():
    p.requires_grad = False

syncnet_T = 5
syncnet_mel_step_size = 16


class Dataset(object):
    def __init__(self, split):
        self.all_videos = get_image_list(args.data_root, split)

    def get_frame_id(self, frame):
        return int(basename(frame).split('.')[0])

    def get_window(self, start_frame):
        start_id = self.get_frame_id(start_frame)
        vidname = dirname(start_frame)

        window_fnames = []
        for frame_id in range(start_id, start_id + syncnet_T):
            frame = join(vidname, '{}.jpg'.format(frame_id))
            if not isfile(frame):
                return None
            window_fnames.append(frame)
        return window_fnames

    def crop_audio_window(self, spec, start_frame):
        # num_frames = (T x hop_size * fps) / sample_rate
        start_frame_num = self.get_frame_id(start_frame)
        start_idx = int(80. * (start_frame_num / float(hparams.fps)))

        end_idx = start_idx + syncnet_mel_step_size

        return spec[start_idx: end_idx, :]

    def __len__(self):
        return len(self.all_videos)

    def __getitem__(self, idx):
        while 1:
            idx = random.randint(0, len(self.all_videos) - 1)
            vidname = self.all_videos[idx]

            img_names = list(glob(join(vidname, '*.jpg')))
            if len(img_names) <= 3 * syncnet_T:
                continue
            img_name = random.choice(img_names)
            wrong_img_name = random.choice(img_names)
            while wrong_img_name == img_name:
                wrong_img_name = random.choice(img_names)

            if random.choice([True, False]):
                y = torch.ones(1).float()
                chosen = img_name
            else:
                y = torch.zeros(1).float()
                chosen = wrong_img_name

            window_fnames = self.get_window(chosen)
            if window_fnames is None:
                continue

            window = []
            all_read = True
            for fname in window_fnames:
                img = cv2.imread(fname)
                if img is None:
                    all_read = False
                    break
                try:
                    img = cv2.resize(img, (hparams.img_size, hparams.img_size))
                except Exception as e:
                    all_read = False
                    break

                window.append(img)

            if not all_read: continue

            try:
                wavpath = join(vidname, "audio.wav")
                wav = audio.load_wav(wavpath, hparams.sample_rate)

                orig_mel = audio.melspectrogram(wav).T
            except Exception as e:
                continue

            mel = self.crop_audio_window(orig_mel.copy(), img_name)

            if (mel.shape[0] != syncnet_mel_step_size):
                continue

            # H x W x 3 * T
            x = np.concatenate(window, axis=2) / 255.
            x = x.transpose(2, 0, 1)
            x = x[:, x.shape[1] // 2:]

            x = torch.FloatTensor(x)
            mel = torch.FloatTensor(mel.T).unsqueeze(0)

            return x, mel, y


ys = {0.: [], 1.: []}


def get_sync_loss(mel, x, y):
    # B, 3 * T, H//2, W
    x = x.to(device)
    mel = mel.to(device)
    y = y.to(device)
    a, v = syncnet(mel, x)
    loss = cosine_loss(a, v, y)
    for i, gt in enumerate(y):
        ys[gt.item()].append(loss[i].item())


def cosine_loss(a, v, y):
    logloss = nn.BCELoss(reduction='none')
    # d = nn.functional.cosine_similarity(a, v, dim=0)
    d = nn.functional.cosine_similarity(a, v)
    # print(d.item(), y.item())
    loss = logloss(d.unsqueeze(1), y)

    return loss


def load_img_mels(path):
    tensors = torch.load(path)
    img_batch = (tensors['img_batch']).detach().cpu().numpy()  # [24, 6, 96, 96]
    mel_batch = tensors['mel_batch'].detach().cpu().numpy()  # [24, 1, 80, 16]

    imgs = img_batch[:5, 3:, :, :]
    mels = mel_batch[:5, :, :, :]
    mel_list = []
    for i in range(mels.shape[0]):
        mel_list.append(mels[i])
    mels = np.concatenate(mel_list, axis=2)

    return torch.FloatTensor(imgs).to(device), torch.FloatTensor(mels).to(device)


def _load(checkpoint_path):
    if use_cuda:
        checkpoint = torch.load(checkpoint_path)
    else:
        checkpoint = torch.load(checkpoint_path,
                                map_location=lambda storage, loc: storage)
    return checkpoint


def load_checkpoint(path, model, optimizer, reset_optimizer=False):
    print("Load checkpoint from: {}".format(path))
    checkpoint = _load(path)
    s = checkpoint["state_dict"]
    new_s = {}
    for k, v in s.items():
        new_s[k.replace('module.', '')] = v
    model.load_state_dict(new_s)
    if not reset_optimizer:
        optimizer_state = checkpoint["optimizer"]
        if optimizer_state is not None:
            print("Load optimizer state from {}".format(path))
            optimizer.load_state_dict(checkpoint["optimizer"])

    return model


if __name__ == "__main__":
    tensors_path = "sync_inputs"
    load_checkpoint(args.syncnet_checkpoint_path, syncnet, None, reset_optimizer=True)

    for i in range(100):
        train_dataset = Dataset('train')
        train_data_loader = data_utils.DataLoader(
            train_dataset, batch_size=2, shuffle=False,
            num_workers=hparams.num_workers)
        for (x, mel, y) in train_data_loader:
            get_sync_loss(mel, x, y)

    y_sync_loss = np.mean(np.asarray(ys[1.]))
    y_unsync_loss = np.mean(np.asarray(ys[0.]))
    print(y_sync_loss, y_unsync_loss)

```

**Comment by burning846** [comment]
I did the same experiment and had the same results. I guess the network is overfitted on the visual images rather than the audios (because of skip-connection I guess) so the audio information was ignored.

**Comment by bredban323** [comment]
Thanks @burning846 for your input. Good to know someone else was able to replicate my results. I am not sure I buy the "overfitted to images due to skip-connections" argument. Although it is plausible, when I fitted the model with L1 loss only for 30k epochs, I can get back the original GT frame by inputting GT/2 and melspectrogram. The only issue with this approach is the lack of temporal consistency when creating the final video using `inference.py`.

This indicates to me that the SyncNet is doing its job however the pretrained weights are highly overfitted to LRS2 dataset. **Does anyone have a video they used for getting SyncNet loss below 0.2?** I have a feeling if this can be done, that video can be reproduced on the output of Wav2Lip.

**Comment by burning846** [comment]
I didn't quite understand "when I fitted the model with L1 loss only for 30k epochs". Do you mean less training can get better results while overfitting can cause damages? 

I tried to train the SyncNet with just one video and used it to train the model. the sync loss could rapidly get down to below 1e-3 while the quality of the output video is very bad. So I think the SyncNet needs a lot of data to train.

**Comment by VERMANs** [comment]
I seem to have completed my training on single-person video footage (training syncnet models to wav2lip models starting with random values), but unfortunately the results are not natural (lips vibrate unnaturally and lips do not close as expected when the background is silent), and the trend towards clarity and synchronization seems to have been achieved.

one of the results are as follows，( this audio.mp3 did not participate in the training
https://user-images.githubusercontent.com/41277638/148172805-a6dcca6d-167d-43cd-a482-bb58c73a93a7.mp4

**Comment by burning846** [comment]
> @burning846 By "when I fitted the model with L1 loss only for 30k epochs", I mean that I used `wav2lip_train.py` however since the sync loss would never average above 0.75 during eval steps, the syncnet_wt would always equal 0. Therefore, the model was trained on L1 loss only. I continued this training for 30k epochs to see if the model has capacity to learn infilling based on melspectrogram and masked GT images + reference. I found that for a given reference + masked GT + melspectrogram, the individual frames do successfully overfit. So if I give those three pieces of information, the generator outputs the GT frame.
> 
> I'm interested in what you said about the SyncNet training on a single video. Were you able to get the SyncNet loss down below 0.2? I think you may be right that SyncNet needs a lot of data to train. Is it possible for you to share the single video that you used to train SyncNet?

My training videos could not be shared but @VERMANs show his result which is almost the same as mine, while my results happened to be a video of the closed mouth with a non-silent clip of voice. My sync loss and L1 loss decrease rapidly to below 1e-3 which I don't think is normal but I didn't find problems in my training code. I guess "why lips do not close as expected when the background is silent" is caused by the lack of silent clips with the closed mouth in training videos, but I didn't have time to prove it so if you have some progress please let me know. Thank you!

**Comment by burning846** [comment]
> I seem to have completed my training on single-person video footage (training syncnet models to wav2lip models starting with random values), but unfortunately the results are not natural (lips vibrate unnaturally and lips do not close as expected when the background is silent), and the trend towards clarity and synchronization seems to have been achieved.
> 
> one of the results are as follows，( this audio.mp3 did not participate in the training https://user-images.githubusercontent.com/41277638/148172805-a6dcca6d-167d-43cd-a482-bb58c73a93a7.mp4

BTW may I ask for the resolution of your training videos? It doesn't seem to be the default resolution 96*96.

**Comment by VERMANs** [comment]
Yeah, I changed the network structure( 288*288). refer to primepake：https://github.com/primepake/wav2lip_288x288

**Comment by xzforeverlove** [comment]
> 我似乎已经完成了对单人视频片段的训练（从随机值开始将同步网络模型训练为 wav2lip 模型），但不幸的是结果并不自然（嘴唇不自然地振动，并且当背景无声时嘴唇不会像预期的那样闭合） ，并且似乎已经实现了清晰和同步的趋势。
> 
> 其中一个结果如下，（这个audio.mp3没有参加训练 https://user-images.githubusercontent.com/41277638/148172805-a6dcca6d-167d-43cd-a482-bb58c73a93a7.mp4
What is the size of the output face, 96*96? Or how much?
